Support multiple providers and regions

akshayknarayan opened the pull request:
Note: this PR changes the tsunami API significantly.

This PR adds support for VM deployments across cloud providers and cloud provider regions. That is, the following is now possible:

```
...
az.add("azure_uswest", ...);
aws.add("ec2_useast", ...);
let az = az.spawn()?;
let aws = aws.spawn()?;
let vms = az.get_machines().extend(aws.get_machines());
...
```

To support this, the existing functionality is moved to src/providers/aws.rs, and put behind a `Launcher` trait, which is a common interface across cloud providers.

An Azure provider is implemented (src/providers/azure.rs), as is one that just wraps a connection to an existing bare-metal machine (src/providers/baremetal.rs).

akshayknarayan committed:
aws: Refactor code to allow multi-region tsunamis

akshayknarayan committed:
aws: Don't timeout ssh connections

akshayknarayan committed:
aws: Clean up sg and key

akshayknarayan committed:
example: Run faster

akshayknarayan committed:
TsunamiBuilder: nickname support

akshayknarayan committed:
ssh: Return stderr and check exit status

akshayknarayan committed:
TsunamiBuilder: clarify documentation

akshayknarayan committed:
aws_region: Security group setup to allow all traffic

TODO: allow for defining security group rules according to the workload
that is going to be run.

akshayknarayan committed:
errors: Nicer error handling

akshayknarayan committed:
run: Add pause support

akshayknarayan committed:
api: Pass logger to closures

akshayknarayan committed:
providers: Cloud provider abstraction

akshayknarayan committed:
baremetal: Add support for baremetal (non-cloud) provider

akshayknarayan committed:
providers: Remove unsafe impl Send with extra trait bound

akshayknarayan committed:
aws: Note down geographic locations of aws regions

akshayknarayan committed:
bin/launch: Utility to quickly launch an aws spot instance

akshayknarayan committed:
ssh: Attempt to support agent-based auth

akshayknarayan committed:
ssh: Save addr field in ssh::Session

akshayknarayan committed:
providers: Add azure provider

akshayknarayan committed:
tsunami: Return machines coupled to provider lifetime

akshayknarayan committed:
providers: Clarify termination behavior for azure, baremetal

akshayknarayan committed:
docs: Fix doctests

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,489 @@
+            .output()?;
+
+        if !out.status.success() {
+            println!("{}", String::from_utf8(out.stdout).unwrap());
```
There are a bunch of `println` around the place that should probably go away :)

jonhoo commented:
This is awesome! I'll sit down and do a review once I have some time to do it thoroughly :)

jonhoo commented on src/lib.rs:
```diff
@@ -53,189 +34,67 @@ extern crate rayon;
-    /// The publicly accessible hostname of this host.
+/// A handle to an instance currently running as part of a tsunami.
+/// Run commands on the machine via [ssh](ssh::Session)
+#[derive(Default)]
```
It's a little weird for this to `derive(Default)`.

jonhoo commented on src/lib.rs:
```diff
@@ -53,189 +34,67 @@ extern crate rayon;
 
-    /// The publicly accessible hostname of this host.
+/// A handle to an instance currently running as part of a tsunami.
+/// Run commands on the machine via [ssh](ssh::Session)
```
Should this link to a particular field on `Machine`? Maybe then use:
```rust
/// Run commands on the machine using the [`ssh::Session`] available through the `ssh` field.
```

jonhoo commented on src/lib.rs:
```diff
@@ -53,189 +34,67 @@ extern crate rayon;
-/// [`TsunamiBuilder#add_set`](struct.TsunamiBuilder.html#method.add_set)) for how to construct a
-/// tsunami.
+/// Call [add](TsunamiBuilder::add) to add machines to the Tsunami, and
+/// [spawn](TsunamiBuidler::spawn) to spawn them and yield a [Tsunami](Tsunami).
```
```suggestion
/// [spawn](TsunamiBuidler::spawn) to spawn them and yield a [`Tsunami`].
```

jonhoo commented on src/lib.rs:
```diff
@@ -53,189 +34,67 @@ extern crate rayon;
-/// spawned as EC2 spot instances. See
-/// [`TsunamiBuilder#add_set`](struct.TsunamiBuilder.html#method.add_set)) for how to construct a
-/// tsunami.
+/// Call [add](TsunamiBuilder::add) to add machines to the Tsunami, and
```
I'd prefer if method calls and such here where surrounded by backticks to give code font.

akshayknarayan replied:
I *think* I got all of them?

jonhoo commented on src/lib.rs:
```diff
@@ -53,189 +34,67 @@ extern crate rayon;
+impl<L: Launcher> TsunamiBuilder<L> {
+    /// Add a machine descriptor to the Tsunami.
+    /// Machine descriptors are specific to the cloud provider they will be used for.
+    pub fn add(&mut self, nickname: String, m: <L as providers::Launcher>::Machine) {
```
I _think_ you can just use `m: L::Machine` here, no?

jonhoo replied:
Also, `nickname` should probably be `&str` here.

jonhoo commented on src/lib.rs:
```diff
@@ -53,189 +34,67 @@ extern crate rayon;
-        self.cluster = false;
+impl<L: Launcher> TsunamiBuilder<L> {
+    /// Add a machine descriptor to the Tsunami.
+    /// Machine descriptors are specific to the cloud provider they will be used for.
```
You probably want to leave an extra newline after the first doc line. It causes rustdoc to realize that the first line is a summary (I believe).

jonhoo commented on src/lib.rs:
```diff
@@ -246,18 +105,17 @@ impl TsunamiBuilder {
-    pub fn set_max_duration(&mut self, hours: u8) {
-        self.max_duration = hours as i64 * 60;
+    /// The default duration is 6 hours.
+    pub fn set_max_duration(&mut self, hours: u64) {
```
You _could_ add a marker trait here along the lines of `SupportsDefinedDuration`, and then here add a bound:
```rust
where L: SupportsDefinedDuration
```
So that people don't accidentally use this where not supported.

jonhoo commented on src/lib.rs:
```diff
@@ -277,532 +135,100 @@ impl TsunamiBuilder {
-        debug!(log, "connecting to ec2");
-
-        let ec2 = rusoto_ec2::Ec2Client::new_with(HttpClient::new()?, provider, self.region);
+    /// Start up all the hosts. Returns a handle, a [Tsunami](Tsunami), from which SSH connections
```
```suggestion
    /// Start up all the hosts. Returns a handle, a [`Tsunami`], from which SSH connections
```

jonhoo commented on src/lib.rs:
```diff
@@ -277,532 +135,100 @@ impl TsunamiBuilder {
+    /// Use to access the machines. The `HashMap` of machines is keyed by the friendly names
+    /// assigned by the call to [add](TsunamiBuilder::add).
+    /// The returned `Machine`s will live for the lifetime of self.
+    pub fn get_machines<'l>(&'l self) -> Result<HashMap<String, Machine<'l>>, Error> {
```
Ah, interesting, so this allows you to "get" the machines multiple times. If you do, you get multiple separate SSH connections to the hosts. Is that right?

jonhoo replied:
If so, may be worth documenting.

jonhoo commented on src/lib.rs:
```diff
@@ -277,532 +135,100 @@ impl TsunamiBuilder {
+/// # Note
+/// See caveats for Azure and Baremetal machines.
+#[must_use]
+pub struct Tsunami<L: Launcher>(Vec<L>, slog::Logger);
```
Probably best to name this fields. Took me a while to realize that the `Vec<L>` is a list of one `Launcher` instance per region.

jonhoo commented on src/lib.rs:
```diff
@@ -277,532 +135,100 @@ impl TsunamiBuilder {
+                HashMap::new,
+                |mut machs, res: Result<HashMap<String, Machine<'l>>, Error>| {
+                    res.and_then(|ms| {
+                        machs.extend(ms);
```
We should warn somewhere about what happens if a user gives two machines the same name.

jonhoo replied:
Or ideally detect this case and error.

akshayknarayan replied:
yep, added a check in `add()` for duplicate keys.

jonhoo commented on src/lib.rs:
```diff
@@ -277,532 +135,100 @@ impl TsunamiBuilder {
+}
 
-        debug!(log, "all done");
+// prevent destructuring
```
Is there a particularly good reason to do so?

akshayknarayan replied:
It's important to prevent the `Vec<L>` in `Tsunami` from escaping right? Otherwise, it could be dropped while the `Machine<'l>`'s returned by `get_machines` were still around.

Or, maybe the borrow checker would catch this?

jonhoo replied:
The borrow checker should catch that.

jonhoo commented on src/providers/mod.rs:
```diff
@@ -0,0 +1,65 @@
+    type Region: Send + Eq + std::hash::Hash + Clone + std::string::ToString;
+    type Machine: MachineSetup<Region = Self::Region> + Send;
+
+    fn init(log: slog::Logger, r: Self::Region) -> Result<Self, Error>;
```
This kind of pattern is pretty weird. I would instead expect that the trait `Launcher` had a method like:
```rust
struct LaunchDescription {
        r: Self::Region
        max_instance_duration: Option<std::time::Duration>,
        max_wait: Option<std::time::Duration>,
        machines: impl IntoIterator<Item = (String, Self::Machine)>,
}

trait Launcher {
    type Machine: Instance;
    fn launch(
        &mut self,
        desc: LaunchDescription
    ) -> HashMap<String, Self::Machine>;
}

trait Instance {
    fn connect(&self) -> Result<crate::Machine, Error>;
}
```
and then that `TsunamiBuilder::spawn` _takes_ an `l: Launcher`.

akshayknarayan replied:
One nice thing about this API is that it auto-groups the `descriptors` into `Self::Region`s, and only initializes the correct number of `impl Launcher`s (see https://github.com/akshayknarayan/tsunami/blob/machine-lifetimes/src/lib.rs#L173).
Passing in an `l: Launcher` would preclude that, and all the callers would have to implement de-duplication like this, which is a little annoying. This is also why `type Region: Hash` is there.

I do think a marker trait like you suggested above might be useful, but these two questions should probably be resolved together.

akshayknarayan replied:
One nice thing about this API is that it auto-groups the `descriptors` into `Self::Region`s, and only initializes the correct number of `impl Launcher`s (see https://github.com/akshayknarayan/tsunami/blob/machine-lifetimes/src/lib.rs#L173).
Passing in an `l: Launcher` would preclude that, and all the callers would have to implement de-duplication like this, which is a little annoying. This is also why `type Region: Hash` is there.

I do think a marker trait like you suggested above might be useful, but these two questions should probably be resolved together.

jonhoo replied:
You can still group them like that inside of `spawn`, since it still knows about each region from the `LaunchDescription`. So if you want them grouped, you can get them grouped.

jonhoo commented on src/ssh.rs:
```diff
@@ -16,16 +15,17 @@ use std::time::{Duration, Instant};
 
 impl Session {
-    pub(crate) fn connect(
+    pub fn connect(
```
Why should this be `pub`?

akshayknarayan replied:
leaving aside tsunami, it's independently useful - making it public makes it possible to use in other crates.

alternately, this module could be its own crate?

jonhoo replied:
Like a "convenient SSH session" crate? That seems like a _great_ idea.

akshayknarayan replied:
https://crates.io/crates/sessh

jonhoo commented on src/ssh.rs:
```diff
@@ -52,17 +53,37 @@ impl Session {
+            sess.userauth_pubkey_file(username, None, key, None)
+                .context("failed to authenticate ssh session with key")?;
+        } else {
+            // TODO doesn't work, see crate::providers::baremetal::test
```
This should probably be fixed or removed.

jonhoo commented on src/ssh.rs:
```diff
@@ -52,17 +53,37 @@ impl Session {
         })
     }
 
     /// Issue the given command and return the command's raw standard output.
```
This should be updated to reflect what it now returns. Maybe it should return a `struct` rather than a tuple?

jonhoo commented on src/ssh.rs:
```diff
@@ -109,15 +130,31 @@ impl Session {
+        }
     }
 
     /// Issue the given command and return the command's standard output.
```
Same here.

akshayknarayan committed:
Machine should not impl Default

akshayknarayan committed:
tsunami: Remove unnecessary impl Drop

akshayknarayan committed:
ssh: Move to standalone crate sessh

akshayknarayan committed:
Tsunami: name the fields

akshayknarayan committed:
doc: Fix documentation

akshayknarayan committed:
TsunamiBuilder::add: Simplify arguments

akshayknarayan committed:
TsunamiBuilder::add: Check for duplicate name

akshayknarayan committed:
get_machines: Document multi-call case

akshayknarayan committed:
TsunamiBuilder: Return &mut Self to allow chained method calls

akshayknarayan committed:
baremetal: Fix default username case

akshayknarayan committed:
tests: Ignore by default

Travis CI doesn't have cloud provider credentials (or ssh-agent access), so can't run the
tests. Use `cargo t -- --ignored` to test locally.

jonhoo commented on src/lib.rs:
```diff
@@ -97,8 +97,14 @@ impl<L: Launcher> TsunamiBuilder<L> {
-        self.descriptors.insert(nickname.to_string(), m);
+    /// They must be unique for each `TsunamiBuilder`. If `nickname` is a duplicate,
+    /// this method will return an `Err` value.
+    pub fn add(&mut self, nickname: &str, m: L::Machine) -> Result<(), Error> {
```
I'd probably just have this return a `bool` instead.

jonhoo commented on src/lib.rs:
```diff
@@ -211,6 +211,10 @@ impl<L: Launcher> Tsunami<L> {
+    ///
+    /// It's possible to call this method multiple times to get multiple
+    /// independent connections to the machines. These will *not* be
+    /// synchronized, so take care to avoid any conflicts.
```
Not clear what "synchronized" means here.

jonhoo commented on src/lib.rs:
```diff
@@ -124,25 +125,29 @@ impl<L: Launcher> TsunamiBuilder<L> {
         let drain = Mutex::new(slog_term::FullFormat::new(decorator).build()).fuse();
         self.log = slog::Logger::root(drain, o!());
+
+        self
```
With this change I think we should make `spawn` take `&self` so that you can use the same builder multiple times if you wish.

jonhoo replied:
See also the [`C-BUILDER` pattern](https://rust-lang-nursery.github.io/api-guidelines/type-safety.html#c-builder).

akshayknarayan replied:
because some of the MachineDescriptors are not `Clone`, it needs to be `&mut self`, with the descriptor list being drained each time.

jonhoo replied:
Why can't `MachineDescriptor` be `Clone`?

akshayknarayan replied:
The descriptors hold a setup function (https://github.com/jonhoo/tsunami/pull/22/files/23e32b6cc39d9a9280c19d016b724e3b9f3bc9a1#diff-60699702750aef5a8cdc050b4fce64eaR58), which was a useful feature in the old API. 
These don't impl Clone?

jonhoo replied:
Hmm, yeah, that's a little awkward. It _is_ possible to bound the function by `Clone`, but you're right that not doing so is _probably_ better.

jonhoo replied:
Although actually, if all you need in `Fn` (i.e., you don't take `FnMut`), then you can just store `Arc` of the setup function, no?

akshayknarayan replied:
Ah, good idea re: `Arc`, it works for the provider implementations we have now.
I could imagine future cases where users might want `FnMut` setup functions, but that can be left to future work.

jonhoo commented on src/providers/baremetal.rs:
```diff
@@ -41,9 +41,12 @@ impl Setup {
-            )?)
+            let user = String::from_utf8(std::process::Command::new("whoami").output()?.stdout)?;
+            let user = user
+                .split_whitespace()
```
This will probably cause issues for people with spaces in their usernames. Perhaps `.lines()` is better? Or perhaps just `.trim()`?

akshayknarayan replied:
aha, I knew I was forgetting the right name for `trim()`...

jonhoo commented on src/providers/aws.rs:
```diff
@@ -739,6 +739,7 @@ mod test {
     use rusoto_ec2::Ec2;
 
     #[test]
+    #[ignore]
```
It'd be awesome to configure Travis so that it had a local SSH endpoint we could connect to (perhaps a container) with a locally generated keypair. That's probably outside the scope of this PR though.

akshayknarayan committed:
Launcher: Remove init() from trait

akshayknarayan committed:
provders: Marker trait SupportsDefinedDuration

akshayknarayan committed:
api: Don't construct Launchers

akshayknarayan committed:
Make the doctest pass

akshayknarayan committed:
TsunamiBuilder: make spawn take &mut self, not self

akshayknarayan committed:
Update documentation

jonhoo commented on examples/test1.rs:
```diff
@@ -16,7 +16,7 @@ fn ping(from: &Machine, to: &Machine, log: &slog::Logger) -> Result<(), failure:
 
 fn main() -> Result<(), failure::Error> {
-    let mut b = TsunamiBuilder::<aws::AWSRegion>::default();
+    let mut b = TsunamiBuilder::<aws::AWSLauncher<_>>::default();
```
I don't think you need this type annotation any more. In fact, I don't think `TsunamiBuilder` needs to be generic over `L` at all. Instead, you can just have `L: Launcher` be a parameter on `spawn`  I think.

akshayknarayan replied:
`TsunamiBuilder` still needs to be`<L: Launcher>` because it internally holds `HashMap<String, L::Machine>`, the provider-specific setup structs.

jonhoo replied:
I'm not sure I follow. Where are those provider-specified entries created?

akshayknarayan replied:
https://github.com/jonhoo/tsunami/pull/22/files/23e32b6cc39d9a9280c19d016b724e3b9f3bc9a1#diff-8a21d79163884c379283e8dcfc728f67R27

The user instantiates an `L::Machine` and passes it to `add(&mut self, nickname: &str, m: L::Machine)`, which is where the type parameter comes in.

I agree the type annotation can be removed, but the type parameter still needs to be there?

jonhoo replied:
I think I would make the builder just generic over the machine type `M` then, and then make `spawn`:
```rust
fn spawn<L: Launcher<Machine = M>>(...)
```

akshayknarayan replied:
Seems reasonable, done

jonhoo commented on examples/test1.rs:
```diff
@@ -30,16 +30,18 @@ fn main() -> Result<(), failure::Error> {
-    let vms = ts.get_machines()?;
-    let log = ts.logger();
+    let mut l: tsunami::providers::aws::AWSLauncher<_> = Default::default();
+    l.with_credentials(|| Ok(DefaultCredentialsProvider::new()?));
```
Should this still call `wait_limit` like before?

akshayknarayan replied:
yes, except `wait_limit` is called `timeout` now.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -154,75 +233,6 @@ pub struct AWSRegion {
-
 impl AWSRegion {
     /// Connect to AWS region `region`, using credentials provider `provider`.
     pub fn new<P>(region: &str, provider: P, log: slog::Logger) -> Result<Self, Error>
```
This probably doesn't need to be `pub` anymore?

akshayknarayan replied:
hm, I sort of like the idea of leaving the lower-level API exposed in case `TsunamiBuilder` doesn't work for some use case. So while it doesn't *need* to be `pub`, there's no harm in making it `pub` - although the pub methods should have documentation explaining what they do.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -628,9 +638,46 @@ impl AWSRegion {
-}
 
-impl super::SupportsDefinedDuration for AWSRegion {}
+    pub fn connect_all<'l>(&'l self) -> Result<HashMap<String, crate::Machine<'l>>, Error> {
```
Shouldn't this method be on `AwsLauncher`?

jonhoo replied:
Does this need to be `pub`?

akshayknarayan committed:
example: Add call to timeout

akshayknarayan committed:
aws: Add documentation for AWSRegion

akshayknarayan committed:
MachineSetup: Use Arc to enable derive(Clone)

This in turn enables TsunamiBuilder::spawn() to take &self instead of
&mut self, and the descriptors need not be drained.

akshayknarayan committed:
TsunamiBuilder: Remove unnecessary type annotations

akshayknarayan committed:
TsunamiBuilder: Generic over MachineSetup, not Launcher

jonhoo commented on src/providers/mod.rs:
```diff
@@ -0,0 +1,76 @@
+
+/// Implement this trait to implement a new cloud provider for Tsunami.
+/// Tsunami will call `launch` once per unique region, as defined by `MachineSetup`.
+pub trait Launcher: Drop + Sized {
```
I think these bounds are unnecessary.

jonhoo commented on src/providers/mod.rs:
```diff
@@ -0,0 +1,76 @@
+/// This is used to group machines into connections
+/// to cloud providers. e.g., for AWS we need a separate
+/// connection to each region.
+pub trait MachineSetup: Send + Clone {
```
I think I'd prefer these on `spawn` rather than directly on the trait. Makes mocking a little easier.

akshayknarayan committed:
provider: Remove bounds from trait and move to individual functions

akshayknarayan committed:
TsunamiBuilder: add_multiple convenience function

add_multiple adds multiple of the same machine descriptor to the
Tsunami.

akshayknarayan committed:
aws: Group spot instance requests

jonhoo commented on src/providers/baremetal.rs:
```diff
@@ -0,0 +1,221 @@
+    }
+}
+
+impl PartialEq for Setup {
```
Why are these impls needed?

akshayknarayan replied:
yep, forgot to remove, they were needed in an old version

jonhoo commented on src/providers/baremetal.rs:
```diff
@@ -0,0 +1,221 @@
+        username: Option<String>,
+    ) -> Result<Self, Error> {
+        let username: Result<String, Error> = username.map(Ok).unwrap_or_else(|| {
+            let user = String::from_utf8(std::process::Command::new("whoami").output()?.stdout)?;
```
I really wish there was a nice way of interacting with `.ssh/config` here :'(

akshayknarayan replied:
Could maybe move this functionality into the `sessh` crate if there was a way to read ssh config like you suggested.

jonhoo commented on src/providers/baremetal.rs:
```diff
@@ -0,0 +1,221 @@
+            username,
+            addr: addr
+                .to_socket_addrs()?
+                .next()
```
I think we may want to keep all the addrs around so that we can try each in turn if need be.

jonhoo commented on src/providers/baremetal.rs:
```diff
@@ -0,0 +1,221 @@
+}
+
+impl super::Launcher for Machine {
+    type Machine = Setup;
```
This reads _really_ weirdly to me. It seems like the `Machine` associated type should maybe have a different name if `Setup` is indeed the right name for the concrete type.

akshayknarayan replied:
Makes sense; I've renamed to `MachineDescriptor`.

jonhoo commented on src/providers/baremetal.rs:
```diff
@@ -0,0 +1,221 @@
+/// The `impl Drop` of this type is a no-op, since Tsunami can't terminate an existing machine.
+#[derive(Default)]
+pub struct Machine {
+    pub log: Option<slog::Logger>,
```
It's pretty tempting to move away from `slog` (which requires you to pass the logger around everywhere) and to something like `tracing` (which doesn't). May be worth looking into. The transition is pretty painless since `tracing` has very similar macros.

akshayknarayan replied:
meh, there are some useful features that passing the logger around enables, like here: https://github.com/jonhoo/tsunami/pull/22/files#diff-b4aea3e418ccdb71239b96952d9cddb6R212

jonhoo replied:
I'm not sure how that link is relevant?

akshayknarayan replied:
Yeah that link seems wrong. I was referring to adding "default fields" to a logger which are currently used to tag loggers with the region they are associated with, etc.

jonhoo replied:
In `tracing` the equivalent is to construct and enter a "`Span`". Converting should probably be done in a separate PR though.

jonhoo commented on src/providers/baremetal.rs:
```diff
@@ -0,0 +1,221 @@
+            .machines
+            .into_iter()
+            .collect::<Vec<(String, Self::Machine)>>();
+        if dscs.is_empty() {
```
I would make this an `assert!`. Trying to launch something that has no machines should be an error in `TsunamiBuilder` already I think.

akshayknarayan replied:
actually, I think `TsunamiBuilder` will happily just not launch anything and return `Ok(())` if you ask it to launch zero machines.
either way, good idea to make this an assert.

akshayknarayan replied:
actually, in the case that this type is called without going through `TsunamiBuilder`, it would retain more flexibility to not panic on this error? So maybe it is better to keep as is.

jonhoo commented on src/providers/baremetal.rs:
```diff
@@ -0,0 +1,221 @@
+        }
+
+        if dscs.len() > 1 {
+            warn!(log, "Discarding duplicate connections to same machine";
```
I think to be helpful to the user this should warn about all the machines that were discarded, rather than the one that wasn't.

jonhoo commented on src/providers/baremetal.rs:
```diff
@@ -0,0 +1,221 @@
+        let dscs = l
+            .machines
+            .into_iter()
+            .collect::<Vec<(String, Self::Machine)>>();
```
You probably don't need to `collect` here. Just keep the iterator, call `next` to get `(name, setup)`, and then `for discarded in rest { warn!(...) }`

jonhoo commented on src/providers/baremetal.rs:
```diff
@@ -0,0 +1,221 @@
+
+        let (name, setup) = dscs.into_iter().next().unwrap();
+
+        let mut sess = ssh::Session::connect(
```
I _think_ we only need to connect here if there is a `setup_fn`?

akshayknarayan replied:
yup

akshayknarayan committed:
aws: Only clean up resources if initialized

akshayknarayan committed:
aws: Add test for launching multiple instances with one spot request

akshayknarayan committed:
baremetal: Remove unnecessary trait impls on Setup

akshayknarayan committed:
baremetal: Only make ssh conn in launch() if setup fn is defined

akshayknarayan committed:
baremetal: Warn about discarded machines

akshayknarayan committed:
Launcher: rename Launcher associated type to MachineDescriptor

akshayknarayan committed:
baremetal: Try all addresses from to_socket_addrs

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+use std::sync::Arc;
+use std::{thread, time};
+
+struct UbuntuAmi(String);
```
This type, or at least its implications for defaults, should be documented somewhere.

akshayknarayan replied:
It is: https://github.com/jonhoo/tsunami/pull/22/files#diff-60699702750aef5a8cdc050b4fce64eaR120

akshayknarayan replied:
Users shouldn't otherwise have to know about this type I think.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+/// A descriptor for a particular machine setup in a tsunami.
+///
+/// The `setup` argument is called once for every spawned instances of this type with a handle
+/// to the target machine. Use [`Machine::ssh`](struct.Machine.html#structfield.ssh) to issue
```
I don't _think_ you need the `()` for the link here?

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+    }
+}
+
+impl PartialEq for MachineSetup {
```
These can probably be removed. And same if the same exists in azure.

akshayknarayan replied:
oops, forgot to do this

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+
+impl Default for MachineSetup {
+    fn default() -> Self {
+        MachineSetup {
```
These defaults should all be documented somewhere.

jonhoo replied:
In fact, does it even make sense to provide a `impl Default` for `MachineSetup`? It strikes me as weird. When would you ever not care?

akshayknarayan replied:
why not :smile:? These defaults are reasonable, and users can replace the ones they want.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+    /// here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#concepts-available-regions)
+    /// AMIs are region-specific. This will overwrite the ami field to
+    /// the Ubuntu 18.04 LTS AMI in the selected region.
+    pub fn region(mut self, region: Region) -> Self {
```
I think the general preference for builder-like types such as this one is to take and return `&mut`. See [`C-BUILDER`](https://rust-lang-nursery.github.io/api-guidelines/type-safety.html#c-builder).

akshayknarayan replied:
Related comment here: https://github.com/jonhoo/tsunami/pull/22/files#diff-60699702750aef5a8cdc050b4fce64eaR306

akshayknarayan replied:
ok, actually, using `&mut` makes this more annoying to use in general, as here: https://github.com/jonhoo/tsunami/pull/22/files#diff-46f9af62cc3650069927935c2a6dd4c8R63

a one-liner with a consuming builder has to get split into multiple lines with `&mut`. The current code is easier to write IMO.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+    /// the Ubuntu 18.04 LTS AMI in the selected region.
+    pub fn region(mut self, region: Region) -> Self {
+        self.region = region.clone();
+        self.ami = UbuntuAmi::from(region).into();
```
It's weird to me for the AMI to change just because I changed region. I guess I see why that'd be the case, but I think it'd be better to simply require that callers provide an AMI when they change regions.

akshayknarayan replied:
This avoids an annoying user problem when they change the region and then get some error about needing to call `ami()`. In most cases, the Ubuntu 18.04 ami is what people want and the region is what they want to change - better to optimize for the common case.

jonhoo replied:
I have never once wanted a stock AMI on EC2 for tsunami experiments, but maybe I'm the odd one out here? I'd personally prefer for us to not have to maintain a list of up-to-date AMIs though. And issuing a breaking change if they ever change. It seems more reasonable to instead just have `region` take a mandatory `ami` argument, no?

akshayknarayan replied:
it's better to have _some_ default (which is documented) than make everyone look up the ami id they want  (maintaining their own version of `UbuntuAmi`) and pass it in?

The only kind of confusing behavior here is something like
```
MachineSetup::default().ami("foobar").region("us-west-2")
```
where the ami gets overwritten. This is unfortunate but documented, and for use cases like the one you said where you are passing an ami id anyway, doesn't get in the way if you are ok with the default region.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+    }
+
+    /// The `setup` argument is called once for every spawned instances of this type with a handle
+    /// to the target machine. Use [`Machine::ssh`](struct.Machine.html#structfield.ssh) to issue
```
I don't think you need the `()` here either.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+
+    /// The `setup` argument is called once for every spawned instances of this type with a handle
+    /// to the target machine. Use [`Machine::ssh`](struct.Machine.html#structfield.ssh) to issue
+    /// commands on the host in question.
```
An example would probably be good here.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+        self
+    }
+
+    /// The `setup` argument is called once for every spawned instances of this type with a handle
```
I think I would phrase this as "the provided callback is called ..."

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+    }
+}
+
+/// Launch AWS EC2 spot instances.
```
```suggestion
/// AWS EC2 spot instance launcher.
```

akshayknarayan replied:
sure

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+/// Launch AWS EC2 spot instances.
+///
+/// This implementation uses [rusoto](https://crates.io/crates/rusoto_core) to connect to AWS.
+/// Therefore, you must call [`with_credentials`](AWSLauncher::with_credentials) to authenticate
```
I don't understand the use of "therefore" here?

akshayknarayan replied:
we use library X, _therefore_ you must provide an authentication method that it understands?

jonhoo replied:
I don't think the library is important. For any library, you'd need to provide some kind of credentials to access the ec2 api?

akshayknarayan replied:
What about: 
```suggestion
/// For it to be useful, you must call [`with_credentials`](AWSLauncher::with_credentials) to authenticate
```

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+/// This implementation uses [rusoto](https://crates.io/crates/rusoto_core) to connect to AWS.
+/// Therefore, you must call [`with_credentials`](AWSLauncher::with_credentials) to authenticate
+/// with AWS.
+/// Each individual region is handled by `AWSRegion`.
```
Does the user ever even see `AWSRegion`?

akshayknarayan replied:
They could if they wanted - goes back to the point about exposing lower-level components for potential unanticipated use

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+impl<P: ProvideAwsCredentials> Default for AWSLauncher<P> {
+    fn default() -> Self {
+        AWSLauncher {
+            credential_provider: None,
```
Wouldn't it make more sense to default this to `DefaultCredentialsProvider`?

jonhoo replied:
I think what I'd do here is define `AWSLauncher` with `P = DefaultCredentialsProvider`, and then `impl<P> Default for AWSLauncher<P> where P: Default`.

akshayknarayan replied:
Doesn't this require specialization?

akshayknarayan replied:
i.e., the comment here: https://github.com/jonhoo/tsunami/pull/22/files/735bcf030958462d79be5bf203c52295d2d0ddfb#diff-60699702750aef5a8cdc050b4fce64eaR190

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+    }
+
+    /// The given AWS EC2 instance type will be used. Note that only [EC2 Defined Duration Spot
+    /// Instance types](https://aws.amazon.com/ec2/spot/pricing/) are allowed.
```
Is that true even when the `max_instance_duration` is `None`?

akshayknarayan replied:
yes - default instance duration if one is not provided is 6 hours.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+/// Therefore, you must call [`with_credentials`](AWSLauncher::with_credentials) to authenticate
+/// with AWS.
+/// Each individual region is handled by `AWSRegion`.
+pub struct AWSLauncher<P: ProvideAwsCredentials> {
```
Do we really need the `ProvideAwsCredentials` bound on the `struct` itself?

akshayknarayan replied:
It's a little annoying, but the credential_provider field, as a closure, needs a return type

jonhoo replied:
Ah, sorry, I'm not proposing removing `P`, just the bound.

akshayknarayan replied:
Why remove the bound - for a P that doesn't follow the bound, the struct is useless?

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+    fn default() -> Self {
+        AWSLauncher {
+            credential_provider: None,
+            max_instance_duration: None,
```
This `None` is a little misleading if we (as the docs below indicate) actually default it to 6 hours?

akshayknarayan replied:
there is an `unwrap_or_else(|| 360)` below

jonhoo replied:
I guess I'm wondering why have it be an `Option` at all?

akshayknarayan replied:
ah, fair enough.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+    }
+}
+
+impl<P: ProvideAwsCredentials> std::ops::Drop for AWSLauncher<P> {
```
Probably not necessary?

akshayknarayan replied:
sure

akshayknarayan committed:
aws: Remove unnecessary impl traits on Setup

akshayknarayan commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,858 @@
+        })
+    }
+
+    fn make_security_group(mut self) -> Result<Self, Error> {
```
new() needs to return Self, and internally it calls two methods that consume and return Self.
These could be made `&mut Self`, but that makes the code in `new()` ugly, something like: 

```
let ec2 = AWSRegion::connect(...)?;
ec2.make_security_group(..)?.make_ssh_key(..)?;
Ok(ec2)
```

rather that the current one-liner.

akshayknarayan committed:
docs: Simplify documentation

akshayknarayan committed:
aws: Remove Option from max_instance_duration

akshayknarayan committed:
aws: Remove Drop impl from AWSLauncher

akshayknarayan committed:
aws: Clearer MachineSetup docs

akshayknarayan committed:
aws: Add example to MachineSetup::setup fn doc

akshayknarayan committed:
aws: Clear up MachineSetup ami setting

jonhoo commented on src/providers/aws.rs:
```diff
@@ -135,15 +135,19 @@ impl MachineSetup {
 pub struct AWSLauncher<P: ProvideAwsCredentials> {
     credential_provider: Option<Box<dyn Fn() -> Result<P, Error>>>,
-    max_instance_duration: Option<std::time::Duration>,
+    max_instance_duration: std::time::Duration,
```
Maybe we should just store this as a number of hours, since it can't be set more granuarly than that anyway?

jonhoo commented on src/providers/aws.rs:
```diff
@@ -48,53 +48,75 @@ impl Into<String> for UbuntuAmi {
 }
 
-impl std::hash::Hash for MachineSetup {
+impl std::hash::Hash for MachineSetup<YesAmi> {
```
Can this impl be removed?

akshayknarayan committed:
aws: Remove unnecessary Hash impl

akshayknarayan committed:
aws: Specify DefaultCredentialsProvider

akshayknarayan committed:
aws: Store max_instance_duration as int representing hours

akshayknarayan committed:
providers: Only establish ssh connections in launch if needed

jonhoo commented on src/providers/aws.rs:
````diff
@@ -220,14 +198,27 @@ where
-    // self.credential_provider.unwrap_or_else(|| Box::new(|| DefaultCredentialProvider::new()))?()
-    // ```
+    /// A closure which returns [`P:
+    /// ProvideAwsCredentials`](https://docs.rs/rusoto_core/0.40.0/rusoto_core/trait.ProvideAwsCredentials.html).
````
This comment doesn't seem quite right. And could also probably be more helpful :p

jonhoo commented on src/providers/aws.rs:
```diff
@@ -169,19 +169,15 @@ impl MachineSetup<YesAmi> {
 pub struct AWSLauncher<P = DefaultCredentialsProvider> {
     credential_provider: Box<dyn Fn() -> Result<P, Error>>,
-    max_instance_duration: std::time::Duration,
+    max_instance_duration_hours: i64,
```
I'd probably just use `isize` here since we don't care too much about its range.

jonhoo replied:
Also, should it be `usize` since the value can't be negative?

jonhoo commented on src/providers/aws.rs:
```diff
@@ -193,8 +189,9 @@ impl<P> AWSLauncher<P> {
     /// maximum). If `t` > 6 hours, `AWSLauncher` will use a duration of 6 hours.
-    pub fn set_max_instance_duration(&mut self, t: std::time::Duration) -> &mut Self {
-        self.max_instance_duration = t;
+    pub fn set_max_instance_duration(&mut self, t: i64) -> &mut Self {
```
This should probably dictate what happens if you set `t = 0`.

akshayknarayan replied:
yup, forgot to add a call to `min`

akshayknarayan replied:
`max`, not `min`

jonhoo commented on src/providers/aws.rs:
```diff
@@ -233,10 +230,7 @@ where
-                360,
-                (self.max_instance_duration.as_secs() as i64 / 60) as i64,
-            ),
+            std::cmp::min(360, self.max_instance_duration_hours * 60),
```
Shouldn't this be `* 3600`?

akshayknarayan replied:
no? x hours * 60 minutes/hour

akshayknarayan replied:
I will add a comment

jonhoo replied:
I thought the API took the duration in seconds?

akshayknarayan replied:
minutes: https://docs.rs/rusoto_ec2/0.41.0/rusoto_ec2/struct.RequestSpotInstancesRequest.html#structfield.block_duration_minutes

akshayknarayan committed:
AwsLauncher: duration as usize and clearer docs

jonhoo commented on src/providers/aws.rs:
```diff
@@ -188,14 +188,16 @@ impl<P> AWSLauncher<P> {
-    /// maximum). If `t` > 6 hours, `AWSLauncher` will use a duration of 6 hours.
-    pub fn set_max_instance_duration(&mut self, t: i64) -> &mut Self {
+    /// maximum). If `t` > 6 hours, `AWSLauncher` will use a duration of 6 hours. If `t` == 0
+    /// hours, `AWSLauncher` will use a duration of 1 hour.
```
I would just say "`t` will be clamped to be between 1 and 6 hours as required by EC2".

jonhoo commented on src/providers/aws.rs:
```diff
@@ -230,7 +232,10 @@ where
         let mut awsregion = AWSRegion::new(&l.region.to_string(), prov, l.log)?;
         awsregion.make_spot_instance_requests(
-            std::cmp::min(360, self.max_instance_duration_hours * 60),
+            std::cmp::min(
```
Do we still need the `min` here?

jonhoo commented on src/providers/aws.rs:
```diff
@@ -188,14 +188,16 @@ impl<P> AWSLauncher<P> {
     }
 
-    /// A closure which returns [`P:
+    /// To override the credential provider, the provided callback should return [`P:
```
> Set the credential provider used to authenticate to EC2.
>
> The provided function is called once for each region, and is expected to produce a [`ProvideAwsCredentials`] that gives access to the region in question.

akshayknarayan committed:
aws: Documentation tweaks, remove redundant checks

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,886 @@
+
+        for (_, reqs) in machines
+            .into_iter()
+            .map(|(name, m)| {
```
I'd like a comment here explaining what this does and why.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,886 @@
+                                self.instances.get_mut(&instance_id).unwrap();
+                            *ipinfo = Some((public_ip.clone(), public_dns.clone()));
+
+                            if let MachineSetup { setup: Some(f), .. } = m_setup {
```
Currently, these setups are all going to be called sequentially I think, which will make setup with many machines _very_ slow.

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,503 @@
+use crate::ssh;
```
I'd like mod-level documentation here that describes what this provider does, and its requirements (for example, it requires that the `az` command be in the user's `$PATH`).

akshayknarayan replied:
Sure, `aws` could probs use module-level docs as well

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,503 @@
+    use serde::{Deserialize, Serialize};
+    use std::process::Command;
+
+    pub fn available_instances_in_region(r: Region) -> Result<Vec<String>, Error> {
```
Should this be `pub(crate)`? It's not `pub`, since the `mod` is not.

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,503 @@
+                "--name",
+                name,
+                "--image",
+                "UbuntuLTS",
```
This should probably be configurable.

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,503 @@
+                "vm",
+                "open-port",
+                "--port",
+                "0-65535",
```
This seems too broad. It should probably only open for traffic *between* the VMs, not all ports externally?

akshayknarayan replied:
The right answer is probably to make which ports to open configurable, but still allow any inbound IPs - my use cases, for example, often have communication from "outside" machines.

jonhoo replied:
I think at the very least we should make sure the AWS and Azure configurations match up.

akshayknarayan replied:
They do match up: https://github.com/akshayknarayan/tsunami/blob/machine-lifetimes/src/providers/aws.rs#L397

jonhoo replied:
Ah, interesting. I feel like my original intention there was to open up only VM-internal communication, but that would rely on VPC, which is still left as a `TODO` there. I guess this is fine for now!

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,503 @@
+    /// To view the available sizes in the relevant region, use:
+    /// `az vm list-sizes -l <region_name>`.
+    pub fn instance_type(self, inst_type: String) -> Result<Self, Error> {
+        if azcmd::available_instances_in_region(self.region)?
```
I don't think the AWS version checks that the instance type is valid this early? Any particular reason you want to do it here rather than in `launch` like AWS does?

akshayknarayan replied:
This is probably a better place to put the check overall - it's more logical that setting a descriptor field causes an error, rather than allowing a bad descriptor field and then failing later.
Azure has an easy api that lists the available types, so why not call it here?

jonhoo replied:
If so, I think we should add the check in the same place for AWS :)
The reason not to do it is that I don't think people expect builder methods to block on a network connection.

akshayknarayan replied:
yeah, the network request is somewhat unfortunate (although I think `az` does some caching).

`az vm create` will fail if you give it a bad instance_type anyway, so might as well delete some code.

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,503 @@
+
+    /// The provided callback (`setup`) is called once for every spawned instances of this type with a handle
+    /// to the target machine. Use [`Machine::ssh`](crate::Machine::ssh) to issue
+    /// commands on the host in question.
```
We should synchronize this text with what is used for the equivalent function in AWS.

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,503 @@
+/// This implementation relies on the [Azure
+/// CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest). 
+///
+/// It also assumes you have previously run `az login` to authenticate with Microsoft.
```
> with *Azure*

?

akshayknarayan replied:
¯\_(ツ)_/¯

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,503 @@
+}
+
+impl AzureRegion {
+    pub fn new(region: &str, log: slog::Logger) -> Result<Self, Error> {
```
Why does this take a `&str` for region, not a `Region`?

akshayknarayan replied:
oops

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,503 @@
+    }
+}
+
+impl std::ops::Drop for AzureLauncher {
```
Should probably be removed.

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,503 @@
+                    &self.resource_group_name,
+                    &vm_name,
+                    &desc.instance_type,
+                    "ubuntu",
```
This should probably also be configurable.

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,503 @@
+
+                azcmd::open_ports(&self.resource_group_name, &vm_name)?;
+
+                if let Setup { setup_fn: Some(f), .. } = desc {
```
This means setup will be run on one machine at a time, which will make setup of many machines very slow.

akshayknarayan committed:
providers: Documentation improvements

akshayknarayan committed:
azure: Make more configurable

jonhoo commented on src/providers/aws.rs:
```diff
@@ -428,16 +454,20 @@ impl AWSRegion {
         for (_, reqs) in machines
             .into_iter()
             .map(|(name, m)| {
+                // attach labels (ami name, instance type)
```
Ideally I'd also like to see this comment explain why those fields are the only ones we need to group by.

jonhoo commented on src/providers/aws.rs:
````diff
@@ -1,3 +1,29 @@
+//! # Example
+//! ```rust,no_run
+//! use rusoto_core::Region;
+//! use slog::info;
````
Why do you `use slog::info` here?

jonhoo commented on src/providers/aws.rs:
````diff
@@ -1,3 +1,29 @@
+//!
+//! # Example
+//! ```rust,no_run
+//! use rusoto_core::Region;
````
I don't think `Region` is used below?

jonhoo commented on src/providers/aws.rs:
```diff
@@ -1,3 +1,29 @@
+//!
+//! let mut b = TsunamiBuilder::default();
+//! b.add("my machine", aws::MachineSetup::default()).unwrap();
+//! let mut l: tsunami::providers::aws::AWSLauncher = Default::default();
```
`aws` is already in scope, so this can just be `aws::AWSLauncher`.

jonhoo replied:
Speaking of, I wonder if `AWSLauncher` should be renamed to `Launcher` given that it's already under `mod aws`?

jonhoo replied:
Also, this might look cleaner as:
```rust
let mut l = aws::AWSLauncher::default();
```

akshayknarayan replied:
yes, and `aws::MachineSetup` should just be `aws::Setup`, to match with the others.
the `AWSLauncher` comment also applies to `AzureLauncher`

jonhoo commented on src/providers/azure.rs:
```diff
@@ -1,9 +1,46 @@
+//! As a result, if your tsunami crashes (i.e., exits without calling `drop()` on [`AzureLauncher`], you must manually terminate your instances
+//! to avoid extra costs.
+//! The easiest way to do this is to delete resource groups beginning with `tsunami_`:
+//! `az group delete --name <name> --yes`.
```
This should probably tell the user how they find resource groups beginning with `tsunami_` first. The text seems to imply that's what it does, but then only shows deletion.

jonhoo commented on src/providers/azure.rs:
````diff
@@ -1,9 +1,46 @@
+//! Also, you must run `az login` to authenticate with Microsoft.
+//!
+//! # Example
+//! ```rust,no_run
````
Same comments for this example as for the AWS one.

akshayknarayan committed:
docs: Simplify examples

akshayknarayan committed:
azure: oops: Make builder take self

akshayknarayan committed:
providers: Regularize + deduplicate struct naming

aws::AWSLauncher -> aws::Launcher
aws::AWSRegion -> aws::RegionLauncher
aws::MachineSetup -> aws::Setup
azure::AzureLauncher -> azure::Launcher
azure::AzureRegion -> azure::RegionLauncher

akshayknarayan committed:
providers: Run setup functions in parallel

Also, de-duplicate between `aws` and `azure` the code that
connects to the vms and runs the setup function.
For `baremetal`, don't do this because it would involve unnecessarily changing the
`SocketAddr` back into a `&str` to call `parse()` in
`providers::setup_machine()`.

akshayknarayan committed:
azure: Remove instance type check in Setup

akshayknarayan committed:
azure: Reorgnaize source to put boring defs at the bottom

akshayknarayan committed:
lib: Add lints

jonhoo commented on src/providers/mod.rs:
```diff
@@ -74,3 +74,50 @@ fn rand_name_sep(prefix: &str, sep: impl Into<Sep>) -> String {
 pub mod azure;
 pub mod baremetal;
+
+fn setup_machine(
```
I think it's fine to re-do the parse of the addr for baremetal given how much code it allows us to reuse. Either way is okay though.

jonhoo commented on src/providers/azure.rs:
```diff
@@ -515,6 +289,232 @@ impl Drop for RegionLauncher {
+///
+/// See https://azure.microsoft.com/en-us/global-infrastructure/locations/ for more information.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum Region {
```
Maybe it'd be an idea to put these in a submodule?

akshayknarayan replied:
I thought about another submodule, but decided against - it's probably not worth the extra source file + indirection.

jonhoo commented on src/lib.rs:
```diff
@@ -37,17 +37,16 @@
+#![warn(missing_copy_implementations)]
+#![warn(trivial_casts)]
+#![warn(trivial_numeric_casts)]
+#![warn(unused_extern_crates)]
```
Also
```rust
#![warn(rust_2018_idioms)]
#![warn(missing_debug_implementations)]
```
We should also probably run `clippy` (if you're not already).

akshayknarayan replied:
There are 4 clippy complaints which I have chosen not to address, all of which are "very complex type used", e.g., 

```
warning: very complex type used. Consider factoring parts into `type` definitions
  --> src/providers/azure.rs:54:9
   |
54 |         Option<Arc<dyn Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync>>,
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#type_complexity
```

jonhoo replied:
Just add
```
#![allow(clippy::type_complexity)]
```
(I think that's the right invocation)

akshayknarayan committed:
lib: Add debug impl lint and missing debug impls

akshayknarayan committed:
lib: Appease clippy on if-let vs .is_some()

akshayknarayan committed:
azure: oops: Use correct username when connecting machines

akshayknarayan committed:
providers: Deduplicate establishing machine connections

jonhoo commented on src/providers/aws.rs:
```diff
@@ -287,12 +293,14 @@ where
     security_group_id: String,
     ssh_key_name: String,
     private_key_path: Option<tempfile::NamedTempFile>,
+    #[educe(Debug(ignore))]
```
Is `Ec2Client` really not `Debug`?!

akshayknarayan replied:
https://docs.rs/rusoto_ec2/0.41.0/rusoto_ec2/struct.Ec2Client.html

nope :(

akshayknarayan committed:
lib: Ignore clippy's complex type complaining

jonhoo commented on README.md:
````diff
@@ -4,47 +4,30 @@
+```rust,no-run
+use tsunami::providers::{aws, Launcher};
+use tsunami::TsunamiBuilder;
+fn main() -> Result<(), failure::Error> {
````
I'd like this example to have some more comments explaining what's going on, more than one machine, include a setup function, etc. This is going to be the _primary_ thing people look at to figure out how to do what they want after all

jonhoo replied:
The old main example is a good case-study:
https://github.com/jonhoo/tsunami/blob/96a2f912f445b57b02a217b4a6998837373989b6/src/lib.rs#L7-L38

jonhoo commented on src/bin/launch.rs:
```diff
@@ -0,0 +1,82 @@
+}
+
+// just launch an instance in the specified region and wait.
+fn main() -> Result<(), failure::Error> {
```
Is this generally a useful tool?

akshayknarayan replied:
Yeah, I have used it sometimes when I just want to launch a vm. Maybe that's just me though :)

jonhoo replied:
How about we move it into `examples/` instead?

jonhoo commented on README.md:
```diff
@@ -4,47 +4,30 @@
+[`TsunamiBuilder`](struct.TsunamiBuilder.html) and [`Tsunami`](struct.Tsunami.html).
 
-## Examples
+# Example
```
In the past, this file has been generated using `cargo readme`. We may want to go away from that, though then I think `README.tpl` should be removed, and we should test the `README` as a doctest using [`doc-comment`](https://docs.rs/doc-comment/0.3.1/doc_comment/).

akshayknarayan replied:
ah, did not know about `cargo readme`. What is `README.tpl`?

jonhoo replied:
It's a template for the readme. Basically gives additional "stuff" beyond just what's in the crate-level doc comment in `src/lib.rs` that should be included in the generated `README.md`.

akshayknarayan replied:
I re-generated the README with `cargo readme` - going away from that can be handled separately I think.

jonhoo commented on src/lib.rs:
```diff
@@ -1,808 +1,268 @@
+//! use tsunami::TsunamiBuilder;
+//! use tsunami::providers::{Launcher, aws};
+//! use rusoto_core::DefaultCredentialsProvider;
+//! fn main() -> Result<(), failure::Error> {
```
Like my comment on the `README` example, I'd like to see more features exercised and explained in comments here, as this is likely the primary place users are going to see "correct" tsunami code.

jonhoo replied:
The old main example is a good case-study:
https://github.com/jonhoo/tsunami/blob/96a2f912f445b57b02a217b4a6998837373989b6/src/lib.rs#L7-L38

jonhoo commented on src/lib.rs:
```diff
@@ -1,808 +1,268 @@
-    setup: Box<Fn(&mut ssh::Session) -> Result<(), Error> + Sync>,
-}
+    /// If `Some(_)`, an established SSH session to this host.
+    #[educe(Debug(ignore))]
```
Wouldn't it be better to just `impl Debug for ssh::Session`?

akshayknarayan replied:
That crate would then run into the same problem: `ssh2::Session` does not `impl Debug`.

jonhoo replied:
That's fine -- just `impl Debug` for it and don't include the session field :p

jonhoo commented on src/lib.rs:
```diff
@@ -1,808 +1,268 @@
+        nickname_prefix: &str,
+        m: M,
+    ) -> Result<&mut Self, Error> {
+        (0..n)
```
I think you can just use a for loop over `std::iter::repeat(m).take(n)` here. Might simplify the code a little.

jonhoo commented on src/lib.rs:
```diff
@@ -1,808 +1,268 @@
+    }
 
-        let ec2 = rusoto_ec2::Ec2Client::new_with(HttpClient::new()?, provider, self.region);
+    /// Start up all the hosts.
```
I think this deserves some more elaborate documentation. In particular in how it relates to the launcher. The fact that this has a `()` return type is likely going to be confusing to users. An example would probably also go a long way.

jonhoo commented on src/lib.rs:
```diff
@@ -1,808 +1,268 @@
+    /// Start up all the hosts.
+    ///
+    /// SSH connections to each instance are accesssible via
+    /// [`connect_all`](providers::Launcher::connect_all).
```
This should probably explicitly call out that this is a method on the launcher.

jonhoo commented on src/lib.rs:
```diff
@@ -1,808 +1,268 @@
-            })
+#[cfg(test)]
+mod test {
+    pub(crate) fn test_logger() -> slog::Logger {
```
What does this test really test?

akshayknarayan replied:
it's a helper method for other tests. Is it supposed to go somewhere else?

jonhoo replied:
Ohhh, I see. I missed that it wasn't `#[test]`. Never mind.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+//! use tsunami::providers::{aws, Launcher};
+//! use tsunami::TsunamiBuilder;
+//!
+//! let mut b = TsunamiBuilder::default();
```
More detail + comments plz. Also, it'd be great to include the one AWS-specific thing we have here (defined durations).

jonhoo replied:
The old main example is a good case-study:
https://github.com/jonhoo/tsunami/blob/96a2f912f445b57b02a217b4a6998837373989b6/src/lib.rs#L7-L38

akshayknarayan replied:
is it canonical to include the same example in multiple places throughout the docs?

jonhoo replied:
Arguably the top-level example should show _mixing_ AWS and Azure tsunamis, but barring that I think it's fine for them to be the same.

akshayknarayan replied:
sure, the mixing idea was relatively straightforward to write.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+        }
+    }
+
+    /// The given AWS EC2 instance type will be used. Note that only [EC2 Defined Duration Spot
```
Missing newline.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+    }
+
+    /// The provided callback, `setup`, is called once for every spawned instances of this type with a handle
+    /// to the target machine. Use [`Machine::ssh`] to issue
```
Weird text wrap.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+    ///
+    /// AMIs are region-specific.
+    /// This will clear the AMI field, which must be set for this struct to be useful.
+    pub fn region(self, region: Region) -> Setup<NoAmi> {
```
I _think_ this method should be callable for `Setup<NoAmi>` as well?

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+    }
+}
+
+/// AWS EC2 spot instance launcher.
```
It'd probably be good to point people towards `TsunamiBuilder::spawn` here, since that's how they're likely to _use_ this type.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+    P: ProvideAwsCredentials + Send + Sync + 'static,
+    <P as ProvideAwsCredentials>::Future: Send,
+{
+    fn get_credential_provider(&self) -> Result<P, Error> {
```
I think this method is only called once. Do we need it to be its own method?

akshayknarayan replied:
Is there any downside to keeping it separate?

jonhoo replied:
Another trait for users/devs to deal with is all

akshayknarayan replied:
I think in the default case, users don't have to think about this trait?

jonhoo replied:
I mean, if you really want to keep it, I guess that's fine. I'm just a fan of getting rid of traits that aren't needed :p

akshayknarayan replied:
fair enough, I removed it :)

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+#[derive(Clone, Copy, Debug)]
+pub struct YesAmi;
+
+/// A descriptor for a particular machine setup in a tsunami.
```
Maybe mention the fact that an AMI has to be set for this type to be used in association with `Launcher`?

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+
+    fn launch(&mut self, l: super::LaunchDescriptor<Self::MachineDescriptor>) -> Result<(), Error> {
+        let prov = self.get_credential_provider()?;
+        let mut awsregion = RegionLauncher::new(&l.region.to_string(), prov, l.log)?;
```
Should the sequence of calls on `RegionLauncher` really just be a method on `RegionLauncher` instead? That makes more sense to me, because that method can then be `pub`, and the various lower-level methods (like `wait_for_spot_instance_requests`) can be non-`pub`.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+            }
+
+            if let Some(wait_limit) = max_wait {
+                if start.elapsed() > wait_limit {
```
I would probably invert this condition and make it `return Ok(())` early so you avoid the extra indentation.

akshayknarayan replied:
it should be a `continue;` I think?

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+                for instance in reservation.instances.unwrap_or_else(Vec::new) {
+                    match instance {
+                        rusoto_ec2::Instance {
+                            state: Some(rusoto_ec2::InstanceState { code: Some(16), .. }),
```
This `16` could use some docs.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+                        log,
+                        &name,
+                        &public_ip,
+                        "ubuntu",
```
This username shouldn't be hard-coded I think?

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+
+        use rayon::prelude::*;
+        self.instances
+            .par_iter()
```
I think we should document somewhere that regions happen sequentially, but instances within a region are set up in parallel.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+                        _tsunami: Default::default(),
+                    };
+
+                    m.connect_ssh(log, "ubuntu", Some(private_key_path.path()))?;
```
This username also probably shouldn't be hard-coded.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+    }
+}
+
+impl std::ops::Drop for RegionLauncher {
```
You should be able to just use `Drop` here.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+}
+
+impl std::ops::Drop for RegionLauncher {
+    fn drop(&mut self) {
```
We should probably describe this dropping behavior somewhere.

akshayknarayan replied:
https://github.com/jonhoo/tsunami/pull/22/commits/9e6f09a2ef9f013e171272b892d80b6ba12bc97b#diff-60699702750aef5a8cdc050b4fce64eaR286

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,519 @@
+//!
+//! Azure does not support Spot or Defined Duration instances.
+//! As a result, if your tsunami crashes (i.e., exits without calling `drop()` on [`Launcher`], you must manually terminate your instances
+//! to avoid extra costs.
```
There's some weird text wrapping going on here.

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,519 @@
+//! use tsunami::providers::{azure, Launcher};
+//! use tsunami::TsunamiBuilder;
+//!
+//! let mut b = TsunamiBuilder::default();
```
Here too I'd like to see a more elaborate example.

akshayknarayan replied:
What do you think the example should do?

jonhoo replied:
See, for example, the old main example:
https://github.com/jonhoo/tsunami/blob/96a2f912f445b57b02a217b4a6998837373989b6/src/lib.rs#L7-L38

akshayknarayan replied:
what about: 
```
//! use tsunami::TsunamiBuilder;
//! use tsunami::providers::{Launcher, aws};
//! use rusoto_core::{DefaultCredentialsProvider, Region};
//! fn main() -> Result<(), failure::Error> {
//!     // Initialize AWS
//!     let mut aws = aws::Launcher::default();
//!
//!     // Initialize a TsunamiBuilder
//!     let mut tb = TsunamiBuilder::default();
//!     tb.use_term_logger();
//!
//!     // Create a machine descriptor and add it to the Tsunami
//!     let m = aws::Setup::default()
//!         .region_with_ubuntu_ami(Region::UsWest1) // default was UsEast1
//!         .setup(|ssh| { // default is a no-op
//!             ssh.run("sudo apt update")?;
//!             ssh.run("curl https://sh.rustup.rs -sSf | sh -- -y")?;
//!         });
//!     tb.add("my_vm", m);
//!
//!     // Launch the VM
//!     tb.spawn(&mut aws)?;
//!
//!     // SSH to the VM and run a command on it
//!     let vms = aws.connect_all()?;
//!     let my_vm = vms.get("my_vm").unwrap();
//!     println!("public ip: {}", my_vm.public_ip);
//!     let ssh = my_vm.ssh.as_ref().unwrap();
//!     ssh.cmd("git clone https://github.com/jonhoo/tsunami")?;
//!     ssh.cmd("cd tsunami && cargo build")?;
//!     Ok(())
//! }
```

jonhoo replied:
I like it!

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,519 @@
+
+impl Setup {
+    /// See also [`Region`].
+    pub fn region(mut self, r: Region) -> Self {
```
Are Azure images the same as AWS images in that they are tied to a region? If so, we'll have to pull the same `HasAmi` trick here.

akshayknarayan replied:
Nope, they are not tied to a region

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,519 @@
+    }
+
+    /// To view the available sizes in the relevant region, use:
+    /// `az vm list-sizes -l <region_name>`.
```
Probably use a code block for that command. Should also specify the default.

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,519 @@
+        self
+    }
+
+    pub fn username(mut self, username: String) -> Self {
```
Docs (shouldn't this produce a warning?)

akshayknarayan replied:
Turns out `missing_docs` was turned off...

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,519 @@
+///
+/// It also assumes you have previously run `az login` to authenticate.
+#[derive(Debug, Default)]
+pub struct Launcher {
```
It'd probably be good to point people towards `TsunamiBuilder::spawn` here, since that's how they're likely to _use_ this type.

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,519 @@
+}
+
+impl Drop for RegionLauncher {
+    fn drop(&mut self) {
```
Same as for AWS, we should probably document this behavior somewhere.

akshayknarayan replied:
It is: https://github.com/jonhoo/tsunami/pull/22/files/6ca63d29e5c6243882bca7d3bb5fca31a53c042b#diff-4fc8e23f1c61397b75b19c2108b2c510R165

I will add for AWS though.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+
+struct UbuntuAmi(String);
+
+impl From<Region> for UbuntuAmi {
```
Maybe move some of this business to the bottom of the file like we did for Azure?

jonhoo commented on src/providers/azure.rs:
```diff
@@ -0,0 +1,519 @@
+                "--admin-username",
+                username,
+                "--ssh-key-value",
+                "@~/.ssh/id_rsa.pub",
```
Should this be hardcoded? Can Azure do ephemeral key generation for the user like AWS can?

jonhoo replied:
For example, I don't have an RSA key...

akshayknarayan replied:
I don't know how to do this ephemerally for Azure - I think the setup process for Azure involves making a key anyway.

jonhoo replied:
Hmm, can we at least make the path configurable then?

akshayknarayan replied:
https://docs.microsoft.com/en-us/azure/virtual-machines/linux/mac-create-ssh-keys#create-an-ssh-key-pair

What we maybe could do is use `--generate-ssh-keys` and `--ssh-dest-key-path` to make a key, then manually delete it in `impl Drop`.

akshayknarayan replied:
https://github.com/akshayknarayan/tsunami/tree/azure-key-gen, but as we discussed offline I can't test it currently.

jonhoo replied:
That looks good to me!

jonhoo commented on src/providers/mod.rs:
```diff
@@ -0,0 +1,124 @@
+pub trait Launcher {
+    type MachineDescriptor: MachineSetup;
+
+    /// Spawn the instances. Implementors should remember enough information to subsequently answer
```
I think a bit more detail may be useful here. For example the fact that this method may be called multiple times.

jonhoo commented on src/providers/mod.rs:
```diff
@@ -0,0 +1,124 @@
+/// to cloud providers. e.g., for AWS we need a separate
+/// connection to each region.
+pub trait MachineSetup {
+    type Region: Eq + std::hash::Hash + Clone + ToString;
```
I'm surprised the linter doesn't complain about lack of docs here? Is the module perhaps not `pub`?

jonhoo commented on src/providers/aws.rs:
```diff
@@ -0,0 +1,887 @@
+        req.ip_protocol = Some("tcp".to_string());
+        req.from_port = Some(0);
+        req.to_port = Some(65535);
+        req.cidr_ip = Some("0.0.0.0/0".to_string());
```
I feel like I commented on this previously, but this is a much wider security policy than tsunami [currently has](https://github.com/jonhoo/tsunami/blob/96a2f912f445b57b02a217b4a6998837373989b6/src/lib.rs#L397). I think maybe we want a flag that dictates whether external traffic should be accepted.

akshayknarayan committed:
TsunamiBuilder: spawn() documentation

akshayknarayan committed:
aws: Documentation improvements

akshayknarayan committed:
aws: Configurable username

akshayknarayan committed:
aws: Remove pub from low-level RegionLauncher methods

akshayknarayan committed:
docs: Add warn(missing_docs) and the missing docs

akshayknarayan committed:
docs: Improve docs

akshayknarayan committed:
aws: Close ports to Internet by default

jonhoo commented on src/providers/aws.rs:
```diff
@@ -397,11 +407,15 @@ impl RegionLauncher {
+        if use_open_ports {
+            req.cidr_ip = Some("0.0.0.0/0".to_string());
+        } else {
+            req.cidr_ip = Some("172.31.0.0/16".to_string());
```
A link to docs for this range would be useful here I think.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -228,6 +230,13 @@ impl<P> Launcher<P> {
 
+    /// The machines spawned on this launcher will have 
+    /// ports open to the public Internet.
+    pub fn open_ports(&mut self) -> &mut Self {
```
I think we want this for azure as well, but since Azure is new it's not a requirement for this PR.

akshayknarayan replied:
I looked briefly at doing the same for azure, but it is somewhat more complicated (I couldn't find an equivalent of `172.31.0.0/16`), so I left it for now.

akshayknarayan committed:
aws: Document 172.31.0.0./16

jonhoo commented:
@akshayknarayan btw, GitHub is being extremely unhelpful and has chosen to hide some open review comments under "Hidden items".

akshayknarayan committed:
TsunamiBuidler: simplify add_multiple

akshayknarayan committed:
docs: Fix main docs example

akshayknarayan committed:
docs: More complicated examples

akshayknarayan committed:
Machine: don't ignore the ssh field, instead use version that impls Debug

akshayknarayan committed:
aws: Inline call to credential_provider

akshayknarayan committed:
readme: Re-generate with cargo readme

akshayknarayan committed:
azure: Generate ephemeral SSH keys

jonhoo commented:
@akshayknarayan I think we got pretty close with this, so would be good to round it off so it can be included in 0.11 :)

akshayknarayan commented:
Agreed - I think the only remaining thing was testing out the ephemeral key generation with azure to make sure it works?

jonhoo commented:
Yeah, I think that's right. And also merge with `master`, which has seen some minor changes.

akshayknarayan committed:
launch: Make an example

akshayknarayan committed:
aws: Change setup logging to debug level

akshayknarayan committed:
Merge branch 'master' into machine-lifetimes

akshayknarayan commented:
@jonhoo does current master compile for you? I get this:
```
error: expected an item keyword
   --> /home/akshay/.cargo/registry/src/github.com-1ecc6299db9ec823/time-0.2.8/src/utc_offset.rs:366:13
    |
366 |             let tm = timestamp_to_tm(datetime.timestamp())?;
    |             ^^^

error: aborting due to previous error

error: could not compile `time`.
```

crate time is a dependency of chrono, which is a dependency of rusoto_signature.

jonhoo commented:
Yeah, seems to work fine for me? `cargo update`?

jonhoo commented:
Ah, you should also mark the `baremetal` test as `#[ignore]`.

akshayknarayan commented:
oops, will do

akshayknarayan committed:
baremetal: Mark test as ignore for travis

akshayknarayan committed:
aws: Launcher impls Drop, not RegionLauncher

drop() cannot call async methods. So, RegionLauncher's drop() becomes
async fn shutdown(), and Launcher impls Drop, calling each of the
RegionLaunchers' shutdown().

akshayknarayan committed:
azure: Add launch test + downgrade default instance to free tier

akshayknarayan committed:
azure: Remove custom key directory

Otherwise, the following happens:
ERROR: Deployment failed. Correlation ID: d5964a8d-a6d6-468b-a0d0-984d8347a1a2. {
  "error": {
    "code": "InvalidParameter",
    "message": "Destination path for SSH public keys is currently limited to its default value /home/ubuntu/.ssh/authorized_keys  due to a known issue in Linux provisioning agent.",
    "target": "linuxConfiguration.ssh.publicKeys.path"
  }
}

akshayknarayan committed:
azure: Clarify key behavior

akshayknarayan committed:
aws: House the tokio runtime in the Launcher

jonhoo commented on src/providers/aws.rs:
```diff
@@ -318,18 +321,38 @@ where
+            self.rt = Some(
+                tokio::runtime::Builder::new()
+                    .basic_scheduler()
+                    .enable_io()
```
I'd probably make this `enable_all`.

jonhoo commented on src/providers/aws.rs:
```diff
@@ -241,6 +241,7 @@ pub struct Launcher<P = DefaultCredentialsProvider> {
     max_instance_duration_hours: usize,
     use_open_ports: bool,
     regions: HashMap<<Setup as super::MachineSetup>::Region, RegionLauncher>,
+    rt: Option<tokio::runtime::Runtime>,
```
Any particular reason to make this an `Option`?

akshayknarayan replied:
Because of the `impl Drop`, it can't be destructured, so the Option makes https://github.com/jonhoo/tsunami/pull/22/files/df53969d0e11c5b97abc9b6cd06dd9a4e8e441da..ccb5f748dea088cddafb6a4b02d24a5cef73e933#diff-60699702750aef5a8cdc050b4fce64eaR297 work. Is there a better way?

jonhoo replied:
Ah, hmm, that's awkward. In its current form, that won't work anyway, since if you call `with_credentials`, the runtime is taken from the old `self`, which is then dropped, and its `Drop` call will then fail the `expect` on `self.rt`.

Come to think of it, why aren't the various configuration options on `Setup` instead of `Launcher`? Seems to me like once you have a `Launcher` all the settings should already have been set up, including credentials? That'd solve our problem I think.

akshayknarayan replied:
I think it is safe (added a code comment with my reasoning)? Also, it seems silly to per-machine define something which naturally applies to the whole launcher.

jonhoo replied:
Ah, sorry, I was thinking about a launcher builder, but that's not what `Setup` is... Hmm.. I guess it's probably fine.

akshayknarayan committed:
aws: Enable_all on tokio runtime

akshayknarayan committed:
aws: Add comment explaining Launcher::drop()

jonhoo commented:
Before merging this, I want to try using it with the noria benchmarking stuff to stress-test it a little. Will get back to you once I've had a chance to do that!

akshayknarayan commented:
Closing in favor of a cleaned-up version in a few minutes.
